mlir/lib/Quantum/Transforms/PropagateSimpleStatesAnalysis.hpp

```diff
@@ -0,0 +1,236 @@
+  public:
+    PropagateSimpleStatesAnalysis(Operation *target)
+    {
+        // `target` is a qnode function
+        target->walk([&](Operation *op) {
+            if (op->getNumResults() != 1) {
+                // restrict to single-qubit gates
+                return;
+            }
+
+            Value res = op->getResult(0);
+            if (!isa<quantum::QubitType>(res.getType())) {
+                // not a qubit value
+                return;
+            }
+
+            // starting qubits are in |0>
+            if (isa<quantum::ExtractOp>(op)) {
+                qubitValues[res] = QubitState::ZERO;
+                return;
+            }
+
+            assert(isa<quantum::CustomOp>(op));
+
+            // takes in parameters other than the parent qubit
+            // e.g. the rotation angle
+            // must be NOT_A_BASIS!
+            if (op->getNumOperands() != 1) {
+                qubitValues[res] = QubitState::NOT_A_BASIS;
+                return;
+            }
+
+            // get state from parent and gate
+            StringRef gate = cast<quantum::CustomOp>(op).getGateName();
+            Value parent = op->getOperand(0);
+            assert(qubitValues.contains(parent));
+
+            // non basis states stay as non basis states
+            if (qubitValues[parent] == QubitState::NOT_A_BASIS) {
+                qubitValues[res] = QubitState::NOT_A_BASIS;
+                return;
+            }
+
+            // Identity preserves parent state
+            if (gate == "Identity") {
+                qubitValues[res] = qubitValues[parent];
+                return;
+            }
+
+            // A valid FSM transition gate
+            // Special treatment for S+ gate from |L> and |R>
+            if (((qubitValues[parent] == QubitState::LEFT) ||
+                 (qubitValues[parent] == QubitState::RIGHT)) &&
+                (gate == "S")) {
+                if (op->hasAttr("adjoint")) {
+                    switch (qubitValues[parent]) {
+                    case QubitState::LEFT:
+                        qubitValues[res] = QubitState::PLUS;
+                        break;
+                    case QubitState::RIGHT:
+                        qubitValues[res] = QubitState::MINUS;
+                        break;
+                    default:
+                        // this will never trigger as the switch is inside an if
+                        break;
+                    }
+                }
+                else {
+                    qubitValues[res] = QubitState::NOT_A_BASIS;
+                }
+                return;
+            }
+
+            // A valid FSM transition gate
+            if (QubitTransitions[qubitValues[parent]].count(gate) == 1) {
+                qubitValues[res] = QubitTransitions[qubitValues[parent]][gate];
+            }
+            // Not a valid FSM transition gate
+            else {
+                qubitValues[res] = QubitState::NOT_A_BASIS;
+            }
```

Comment:
Before starting the analysis, I would make sure that all gates are legal.

Reply:
Resolving as it duplicates the comment below